src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h

```diff
@@ -0,0 +1,269 @@
+    tune();
+  }
+
+  inline bool tune(bool currentInvalid = false) override;
+
+  inline std::set<ContainerOption> getAllowedContainerOptions() const override { return _containerOptionsSet; }
+
+  inline bool searchSpaceIsTrivial() const override;
+
+  inline bool searchSpaceIsEmpty() const override;
+
+ private:
+  /**
+   * Generate n samples and predict their corresponding
+   * acquisition function. The result is sorted depending on
+   * given acquistion function.
+   * @param n numSamples
+   * @param af acquisition function
+   */
+  inline void sampleAcquisitions(size_t n, AcquisitionFunctionOption af);
+
+  std::set<ContainerOption> _containerOptionsSet;
+  std::vector<ContainerOption> _containerOptions;
+  std::vector<TraversalOption> _traversalOptions;
+  std::vector<DataLayoutOption> _dataLayoutOptions;
+  std::vector<Newton3Option> _newton3Options;
+  std::unique_ptr<NumberSet<double>> _cellSizeFactors;
+
+  std::map<TraversalOption, ContainerOption> _traversalContainerMap;
+
+  FeatureVector _currentConfig;
+  std::vector<GaussianCluster::VectorAcquisition> _currentAcquisitions;
+  std::unordered_set<FeatureVector, ConfigHash> _invalidConfigs;
+
+  Random _rng;
+  std::vector<int> _dimRestrictions;
+  GaussianCluster _gaussianCluster;
+  size_t _maxEvidence;
+  AcquisitionFunctionOption _predAcqFunction;
+  size_t _predNumLHSamples;
+};
+
+bool BayesianClusterSearch::tune(bool currentInvalid) {
+  if (currentInvalid) {
+    _invalidConfigs.insert(_currentConfig);
+  }
+
+  if (searchSpaceIsEmpty()) {
+    // no valid configuration
+    _currentConfig = FeatureVector();
+    return false;
+  }
+
+  // no more tunings steps
+  if (_gaussianCluster.numEvidence() >= _maxEvidence) {
+    // select best config
+    _currentConfig = FeatureVector::clusterDecode(_gaussianCluster.getEvidenceMin(), _traversalOptions,
+                                                  _dataLayoutOptions, _newton3Options);
+    _currentConfig.container = _traversalContainerMap[_currentConfig.traversal];
+    AutoPasLog(debug, "Selected Configuration {}", _currentConfig.toString());
+    return false;
+  }
+
+  // try to sample a valid vector which is expected to yield a good acquisition
+  for (size_t i = 0; i < maxAttempts; ++i) {
+    if (_currentAcquisitions.empty()) {
+      sampleAcquisitions(_predNumLHSamples, _predAcqFunction);
+    }
+
+    // test best vectors until empty
+    while (not _currentAcquisitions.empty()) {
+      auto best = FeatureVector::clusterDecode(_currentAcquisitions[0].first, _traversalOptions, _dataLayoutOptions,
+                                               _newton3Options);
+      best.container = _traversalContainerMap[best.traversal];
+
+      if (_invalidConfigs.find(best) == _invalidConfigs.end()) {
+        // valid config found!
+        _currentConfig = best;
+        return true;
+      } else {
+        // invalid: dispose
+        _currentAcquisitions.erase(_currentAcquisitions.begin());
```

Comment:
It seems that the only operations you do with `_currentAcquisitions` are to delete the first element, sorting and accessing the first. Maybe a [`std::deque`](https://en.cppreference.com/w/cpp/container/deque) might be better than a `std::vector`? Popping the front of a vector is neither cheap nor nice.